examples/test.py

```diff
@@ -0,0 +1,117 @@
+import hydra
+
+from omegaconf import DictConfig
+from omegaconf import OmegaConf
+from typing import List
+
+from enum import Enum
+from dataclasses import dataclass, field
+
+#@dataclass
+class NeMoConfig(dict):
+    """ Abstract NeMo Configuration class"""
+    pass
+    def serialize(self):
+        """ Maybe we will have it in here - or maybe not ;) """
+        pass
+
+    def deserialize(self):
+        """ Maybe we will have it in here - or maybe not ;) """
+        pass
+
+class Height(Enum):
+    """ Enum - just to test how enums can be nested """
+    SHORT = 0
+    TALL = 1
+
+@dataclass
+class ModuleConfig(NeMoConfig):
+    """ This is a module config  - just to test the inheritance """
+    name: str="my_name"
+    val: int=1
+    height: Height = Height.SHORT # Enum!
+
+@dataclass
+class OptimConfig(NeMoConfig):
+    """ This is an optimizer config """
+    name: str="sgd"
+    lr: float=0.99
+
+
+@dataclass
+class JasperBlockConfig(NeMoConfig):
+    """ Default config of a single Jasper block """
+    dilation: int=1
+    dropout: float=0.0
+    filters: int=256
+    kernel: int=33
+    repeat: int=5
+    residual: bool=True
+    separable: bool=True
+    stride: int=1
+
+
+
+@dataclass
+class JasperEncoderConfig(ModuleConfig):
+    """ Config of Jasper encoder """
+    activation: str="relu"
+    conv_mask: bool=True
+    feat_in: int=64
+
+    def __init__(self, **kwargs):
+        """ Init - adds additional parameters, currently - adds 10 JasperBlocks """
+        super().__init__(self, kwargs)
+        # Add blocks.
+        self.blocks = []
+        # Add first block with some custom params.
+        self.blocks.append(JasperBlockConfig(repeat=1, residual=False, stride=2))
+
+        # Add remaining 9 using default params.
+        for i in range(9):
+            self.blocks.append(JasperBlockConfig())
+
+
+@dataclass
+class ModelConfig(NeMoConfig):
+    """ This is a module config  - just to test the inheritance """
+    name: str="my_name"
+
+@dataclass
+class JasperConfig(ModelConfig):
```

Comment:
I can't say I am a fan of having to create a Configuration class for each configuration.
It seems redundant to me to have an actual NeMoModel class (Jasper), a Configuration class (JasperConfig), and a yaml file (Jasper.yaml).

Reply:
One thing that I really like about this is that, these classes are basically replacing the argparser that you'd normally see in your python script. Except! Now we have code completion!

Reply:
@ericharper thanks, this is exactly the goal here.

@blisc I understand your concerns:
1) The way you should think about those config classes is a really "structure", like a "dictionary on steroids"
https://docs.python.org/3/library/dataclasses.html
I am still learning how to use/write them properly, but there are several nice features
2) I need to make sure that it is doable - but honestly I want to go all in, i.e. configuration objects will be the default solution, user won't write yaml files at all (so Jasper + JasperConfig), eventually there will be an option to serialize/deserialize to yaml (use-case: store configuration on NGC)

Reply:
It feels like we are duplicating code in a sense since we have to typehint our class arguments as well as define a configuration object that contains the same information. Is there a way to avoid this duplication?

Reply:
@blisc I agree, honestly I am not satisfied with this solution too..

However, I see two advantages:
 - every JasperBlock comes with default settings, so one need to override the defaults only when required
 - we can incorporate here the custom logic for `dropout` and when e.g. user will set dropout>=0  (e.g. default: -1) for Jasper it will pass it to all blocks (re: @borisgin)

The other option that I see is to have those loooong config files....

Any other ideas?